src/actions/index.js

```diff
@@ -28,14 +28,27 @@ export default config => {
   } = config
 
   let add = async (parentPath, node, { index } = {}) => {
-    node = initObject(node)
-    Tree.walk((node, index, [parent = {}]) => {
-      let path = [...(parent.path || parentPath), node.key]
-      initNode(node, path, extend, types)
-      flat[encode(path)] = node
-    })(node)
-
     let target = getNode(parentPath)
+    // initialize uniqueString cache for the parent of the node to be added here,
+    // since it's not visited during the tree walk
+    let parentDedupeChildren = F.uniqueString(_.map('key', target.children))
+    node = initObject(node)
+    Tree.walk(
```

Comment:
The bulk of this walk is repeated here and in the index. We should probably factor something out for this